File: FAST2/Models/ServerCollection.vb
Imports System.Xml.Serialization

Namespace Models

    <Serializable()>
    Public Class ServerCollection
        <XmlElement(Order:=1)>
        Public Property CollectionName As String = "Main"

        <XmlElement(Order:=2, ElementName:="ServerProfile")>
        Public ServerProfiles As List(Of ServerProfile) = New List(Of ServerProfile)()

        Private Shared Function GetServerProfiles() As ServerCollection
            Dim currentProfiles As New ServerCollection
            
            If My.Settings.Servers IsNot Nothing
                currentProfiles = My.Settings.Servers
            End If

            Return currentProfiles
        End Function
        
        Public Shared Sub AddServerProfile(name As String, safeName As String)
            Dim duplicate = False
            Dim currentProfiles = GetServerProfiles()
            
            If currentProfiles.ServerProfiles.Count > 0 Then
                For Each profile In currentProfiles.ServerProfiles
                    If profile.DisplayName = name
                        duplicate = true
                    End If
                Next
            End If
           
            If Not duplicate Then
                currentProfiles.ServerProfiles.Add(New ServerProfile(name, safeName))
                My.Settings.Servers = currentProfiles
            Else
                MainWindow.Instance.IMessageDialog.IsOpen = True
                MainWindow.Instance.IMessageDialogText.Text = "Profile Already Exists"
            End If

            My.Settings.Save()
            MainWindow.Instance.LoadServerProfiles
        End Sub

        Public Shared Function RenameServerProfile(oldName As String, newName As String) As Boolean
            Try
                Dim currentProfiles = My.Settings.Servers.ServerProfiles
        
                Dim currentProfile As ServerProfile = currentProfiles.Find(Function(profile) profile.DisplayName = oldName)

                currentProfile.DisplayName = newName
                currentProfile.SafeName = MainWindow.SafeName(newName)
                Return True
            Catch ex As Exception
                MsgBox(ex.Message)
                Return False
            End Try
        End Function

        Public Shared Sub DeleteServerProfile(safeName As String)
            Dim currentProfiles = GetServerProfiles()

            currentProfiles.ServerProfiles.RemoveAll(Function(x) x.SafeName = safeName)

            My.Settings.Save()
            MainWindow.Instance.LoadServerProfiles
        End Sub
    End Class

    <Serializable()>
    Public Class ServerProfile
        Private Sub New()
        End Sub

        Public Sub New(name As String, safeName As String)
            DisplayName =  name
            Me.SafeName = safeName
        End Sub

        Public Property SafeName As String = String.Empty
        Public Property DisplayName As String = String.Empty
        Public Property ServerName As String = String.Empty
        Public Property Executable As String = String.Empty
        Public Property Password As String = String.Empty
        Public Property AdminPassword As String = String.Empty
        Public Property MaxPlayers As Integer = 32
        Public Property Port As Integer = 2302
        Public Property HeadlessClientEnabled As Boolean = False
        Public Property HeadlessIps As String = String.Empty
        Public Property LocalClients As String = String.Empty
        Public Property NoOfHeadlessClients As Integer = 0
        Public Property Loopback As Boolean = False
        Public Property Upnp As Boolean = False
        Public Property Netlog As Boolean = False
        Public Property AutoRestartEnabled As Boolean = False
        Public Property DailyRestartAEnabled As Boolean = False
        Public Property DailyRestartA As Date
        Public Property DailyRestartBEnabled As Boolean = False
        Public Property DailyRestartB As Date
        Public Property VotingEnabled As Boolean = True
        Public Property VotingMinPlayers As Integer = 3
        Public Property VotingThreshold As Decimal = 0.33
        Public Property AllowFilePatching As Integer = 0
        Public Property VerifySignatures As Integer = 0
        Public Property RequiredBuildEnabled As Boolean = False
        Public Property RequiredBuild As Boolean = False
        Public Property KickDuplicates As Boolean = False
        Public Property VonEnabled As Boolean = True
        Public Property CodecQuality As Integer = 3
        Public Property ServerConsoleLogEnabled As Boolean = False
        Public Property ServerConsoleLog As String = "server_console.log"
        Public Property PidEnabled As Boolean = False
        Public Property PidLog As String = "pid.log"
        Public Property RankingEnabled As Boolean = False
        Public Property RankingLog As String = "ranking.log"
        Public Property RptTimestamp As String = "none"
        Public Property Motd As String = String.Empty
        Public Property MotdDelay As Integer = 5
        Public Property ServerMods As String = String.Empty
        Public Property ClientMods As String = String.Empty
        Public Property HeadlessMods As String = String.Empty
        Public Property Missions As String = String.Empty
        Public Property PersistentBattlefield As Boolean = False
        Public Property AutoInit As Boolean = False
        Public Property DifficultyPreset As String = String.Empty
        Public Property ReducedDamage As Boolean = False
        Public Property GroupIndicators As Boolean = False
        Public Property FriendlyNameTags As Boolean = False
        Public Property EnemyNameTags As Boolean = False
        Public Property DetectedMines As Boolean = False
        Public Property MultipleSaves As Boolean = False
        Public Property ThirdPerson As Boolean = False
        Public Property WeaponInfo As Boolean = False
        Public Property StanceIndicator As Boolean = False
        Public Property StaminaBar As Boolean = False
        Public Property CameraShake As Boolean = False
        Public Property VisualAids As Boolean = False
        Public Property ExtendedMapContent As Boolean = False
        Public Property Commands As Boolean = False
        Public Property VonId As Boolean = False
        Public Property KilledBy As Boolean = False
        Public Property Waypoints As Boolean = False
        Public Property Crosshair As Boolean = False
        Public Property AutoReporting As Boolean = False
        Public Property ScoreTable As Boolean = False
        Public Property TacticalPing As Boolean = False
        Public Property AiAccuracy As Decimal = 0.95
        Public Property AiSkill As Decimal = 0.55
        Public Property AiPreset As Integer = 3
        Public Property MaxPacketLossEnabled As Boolean = False
        Public Property MaxPacketLoss As Integer = 0
        Public Property DisconnectTimeoutEnabled As Boolean = False
        Public Property DisconnectTimeout As Integer = 90
        Public Property KickOnSlowNetworkEnabled As Boolean = False
        Public Property KickOnSlowNetwork As String = String.Empty
        Public Property TerrainGrid As Integer = 10
        Public Property ViewDistance As Integer = 2500
        Public Property MaxPingEnabled As Boolean = False
        Public Property MaxPing As Integer = 60
        Public Property MaxDesyncEnabled As Boolean = False
        Public Property MaxDesync As Integer = 0
        Public Property MaxCustomFileSize As Integer = 160
        Public Property MaxPacketSize As Integer = 1400
        Public Property MinBandwidth As Integer = 128
        Public Property MaxBandwidth As Integer = 2000
        Public Property MaxMessagesSend As Integer = 128
        Public Property MaxSizeNonguaranteed As Integer = 256
        Public Property MaxSizeGuaranteed As Integer = 512
        Public Property MinErrorToSend As Decimal = 0.001
        Public Property MinErrorToSendNear As Decimal = 0.01
        Public Property ExtraParams As String = String.Empty
        Public Property EnableHyperThreading As Boolean = False
        Public Property FilePatching As Boolean = False
        Public Property ServerCommandPassword As String = String.Empty
        Public Property DoubleIdDetected As String = String.Empty
        Public Property OnUserConnected As String = String.Empty
        Public Property OnUserDisconnected As String = String.Empty
        Public Property OnHackedData As String = String.Empty
        Public Property OnDifferentData As String = String.Empty
        Public Property OnUnsignedData As String = String.Empty
        Public Property RegularCheck As String = String.Empty
        Public Property BattleEye As Boolean = True
    End Class
End NameSpace

File: FASTER/Models/ServerCollection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Xml.Serialization;

namespace FASTER.Models
{
    [Serializable]
    public class ServerCollection
    {
        [XmlElement(Order = 1)]
        public string CollectionName { get; set; } = "Main";

        [XmlElement(Order = 2, ElementName = "ServerProfile")]
        public List<ServerProfile> ServerProfiles = new List<ServerProfile>();

        private static ServerCollection GetServerProfiles()
        {
            ServerCollection currentProfiles = new ServerCollection();

            if (Properties.Settings.Default.Servers != null)
                currentProfiles = Properties.Settings.Default.Servers;

            return currentProfiles;
        }

        public static void AddServerProfile(string name, string safeName)
        {
            var duplicate = false;
            var currentProfiles = GetServerProfiles();

            if (currentProfiles.ServerProfiles.Count > 0)
            {
                foreach (var profile in currentProfiles.ServerProfiles
                                                       .Where(profile => profile.DisplayName == name)) 
                { duplicate = true; }
            }

            if (!duplicate)
            {
                currentProfiles.ServerProfiles.Add(new ServerProfile(name, safeName));
                Properties.Settings.Default.Servers = currentProfiles;
                ServerProfile profile = Properties.Settings.Default.Servers.ServerProfiles.Find(newProfile => newProfile.SafeName == safeName);
                profile.ServerName = name;
                profile.Executable = Properties.Settings.Default.serverPath + @"\arma3server_x64.exe";
            }
            else
            {
                MainWindow.Instance.IMessageDialog.IsOpen = true;
                MainWindow.Instance.IMessageDialogText.Text = "Profile Already Exists";
            }

            Properties.Settings.Default.Save();
            MainWindow.Instance.LoadServerProfiles();
        }

        public static void AddServerProfile(ServerProfile newProfile)
        {
            var duplicate       = false;
            var currentProfiles = GetServerProfiles();

            if (currentProfiles.ServerProfiles.Count > 0)
            {
                foreach (var profile in currentProfiles.ServerProfiles
                                                       .Where(profile => profile.DisplayName == newProfile.DisplayName)) 
                { duplicate = true; }
            }

            if (!duplicate)
            {
                currentProfiles.ServerProfiles.Add(newProfile);
                Properties.Settings.Default.Servers = currentProfiles;
                ServerProfile profile = Properties.Settings.Default.Servers.ServerProfiles.Find(findProfile => findProfile.SafeName == newProfile.SafeName);
                profile.ServerName = newProfile.ServerName;
                profile.Executable = Properties.Settings.Default.serverPath + @"\arma3server_x64.exe";
            }
            else
            {
                MainWindow.Instance.IMessageDialog.IsOpen   = true;
                MainWindow.Instance.IMessageDialogText.Text = "Profile Already Exists";
            }

            Properties.Settings.Default.Save();
            MainWindow.Instance.LoadServerProfiles();
        }

        public static bool RenameServerProfile(string oldName, string newName)
        {
            try
            {
                var currentProfiles = Properties.Settings.Default.Servers.ServerProfiles;

                ServerProfile currentProfile = currentProfiles.Find(profile => profile.DisplayName == oldName);

                currentProfile.DisplayName = newName;
                currentProfile.SafeName = Functions.SafeName(newName);
                return true;
            }
            catch (Exception ex)
            {
                MessageBox.Show($"ServerConfig - An exception occurred:\n{ex.Message}", "Error");
                return false;
            }
        }

        public static void DeleteServerProfile(string safeName)
        {
            var currentProfiles = GetServerProfiles();

            currentProfiles.ServerProfiles.RemoveAll(x => x.SafeName == safeName);

            Properties.Settings.Default.Save();
        }
    }

    [Serializable]
    public class ServerProfile
    {
        public ServerProfile()
        {}

        public ServerProfile(string name, string safeName)
        {
            DisplayName = name;
            SafeName = safeName;
        }

        public string   SafeName              { get; set; } = string.Empty;
        public string   DisplayName           { get; set; } = string.Empty;
        public string   ServerName            { get; set; } = string.Empty;
        public string   Executable            { get; set; } = string.Empty;
        public string   Password              { get; set; } = string.Empty;
        public string   AdminPassword         { get; set; } = string.Empty;
        public int      MaxPlayers            { get; set; } = 32;
        public int      Port                  { get; set; } = 2302;
        public bool     HeadlessClientEnabled { get; set; }
        public string   HeadlessIps           { get; set; } = "127.0.0.1";
        public string   LocalClients          { get; set; } = "127.0.0.1";
        public int      NoOfHeadlessClients   { get; set; }
        public bool     Loopback              { get; set; }
        public bool     Upnp                  { get; set; }
        public bool     Netlog                { get; set; }
        public bool     AutoRestartEnabled    { get; set; } = false;
        public bool     DailyRestartAEnabled  { get; set; } = false;
        public DateTime DailyRestartA         { get; set; }
        public bool     DailyRestartBEnabled  { get; set; } = false;
        public DateTime DailyRestartB         { get; set; }
        public bool     VotingEnabled         { get; set; } = true;
        public int      VotingMinPlayers      { get; set; } = 3;
        public decimal  VotingThreshold       { get; set; } = 33;
        public int      AllowFilePatching     { get; set; }
        public int      VerifySignatures      { get; set; }
        public bool     RequiredBuildEnabled  { get; set; }
        public bool     RequiredBuild         { get; set; } = false;
        public bool     KickDuplicates        { get; set; }
        public bool     VonEnabled            { get; set; } = true;
        public int      CodecQuality          { get; set; } = 3;

        public bool ServerConsoleLogEnabled { get; set; }

        // Public Property ServerConsoleLog As String = "server_console.log"
        public bool PidEnabled { get; set; }

        // Public Property PidLog As String = "pid.log"
        public bool RankingEnabled { get; set; }

        // Public Property RankingLog As String = "ranking.log"
        public string RptTimestamp             { get; set; } = "none";
        public string Motd                     { get; set; } = string.Empty;
        public int    MotdDelay                { get; set; } = 5;
        public string ServerMods               { get; set; } = string.Empty;
        public string ClientMods               { get; set; } = string.Empty;
        public string HeadlessMods             { get; set; } = string.Empty;
        public bool   ManualMissions           { get; set; }
        public string MissionsClass            { get; set; } = string.Empty;
        public string Missions                 { get; set; } = string.Empty;
        public bool   PersistentBattlefield    { get; set; }
        public bool   AutoInit                 { get; set; }
        public string DifficultyPreset         { get; set; } = "Regular";
        public bool   ReducedDamage            { get; set; }
        public string GroupIndicators          { get; set; } = "Never";
        public string FriendlyNameTags         { get; set; } = "Never";
        public string EnemyNameTags            { get; set; } = "Never";
        public string DetectedMines            { get; set; } = "Never";
        public bool   MultipleSaves            { get; set; }
        public bool   ThirdPerson              { get; set; }
        public string WeaponInfo               { get; set; } = "Never";
        public string StanceIndicator          { get; set; } = "Never";
        public bool   StaminaBar               { get; set; }
        public bool   CameraShake              { get; set; }
        public bool   VisualAids               { get; set; }
        public bool   MapContentFriendly       { get; set; }
        public bool   MapContentEnemy          { get; set; }
        public bool   MapContentMines          { get; set; }
        public string Commands                 { get; set; } = "Never";
        public bool   VonId                    { get; set; }
        public bool   KilledBy                 { get; set; }
        public string Waypoints                { get; set; } = "Never";
        public bool   Crosshair                { get; set; }
        public bool   AutoReporting            { get; set; }
        public bool   ScoreTable               { get; set; }
        public bool   TacticalPing             { get; set; }
        public double AiAccuracy               { get; set; } = 0.95;
        public double AiSkill                  { get; set; } = 0.55;
        public int    AiPreset                 { get; set; } = 3;
        public bool   MaxPacketLossEnabled     { get; set; }
        public int    MaxPacketLoss            { get; set; }
        public bool   DisconnectTimeoutEnabled { get; set; }
        public int    DisconnectTimeout        { get; set; } = 90;
        public bool   KickOnSlowNetworkEnabled { get; set; }
        public string KickOnSlowNetwork        { get; set; } = string.Empty;
        public int    TerrainGrid              { get; set; } = 10;
        public int    ViewDistance             { get; set; } = 2500;
        public bool   MaxPingEnabled           { get; set; }
        public int    MaxPing                  { get; set; } = 60;
        public bool   MaxDesyncEnabled         { get; set; }
        public int    MaxDesync                { get; set; }
        public int    MaxCustomFileSize        { get; set; } = 160;
        public int    MaxPacketSize            { get; set; } = 1400;
        public double MinBandwidth             { get; set; } = 131072;
        public double MaxBandwidth             { get; set; } = 10000000000;
        public int    MaxMessagesSend          { get; set; } = 128;
        public int    MaxSizeNonguaranteed     { get; set; } = 256;
        public int    MaxSizeGuaranteed        { get; set; } = 512;
        public double MinErrorToSend           { get; set; } = 0.001;
        public double MinErrorToSendNear       { get; set; } = 0.01;
        public string CpuCount                 { get; set; } = string.Empty;
        public string MaxMem                   { get; set; } = string.Empty;
        public string ExtraParams              { get; set; } = string.Empty;
        public string AdminUids                { get; set; } = string.Empty;
        public bool   EnableHyperThreading     { get; set; }
        public bool   FilePatching             { get; set; }
        public string ServerCommandPassword    { get; set; } = string.Empty;
        public string DoubleIdDetected         { get; set; } = string.Empty;
        public string OnUserConnected          { get; set; } = string.Empty;
        public string OnUserDisconnected       { get; set; } = string.Empty;
        public string OnHackedData             { get; set; } = string.Empty;
        public string OnDifferentData          { get; set; } = string.Empty;
        public string OnUnsignedData           { get; set; } = string.Empty;
        public string RegularCheck             { get; set; } = string.Empty;
        public bool   BattleEye                { get; set; } = true;
        public bool   enableAdditionalParams   { get; set; } = false;
        public string additionalParams         { get; set; } = string.Empty;
    }
}
